### PR TITLE
test coverage for katello-cert-check invalid/valid input

### PR DESCRIPTION
**Pull request by omkarkhatavkar**
### Description
This PR will be responsible to add new checks for invalid/valid conditions in katello-cert-check command. Add cleanup for certs so any modification can be done on the certs to check more invalid test cases, that will be introduced in further PR's.

If we dont have the cleanup, usually tests can not be re-run and failing with existing files.     

### Test Result 
```
(env) /home/okhatavk/Satellite/robottelo (master) $ pytest --picked --mode=onlychanged

Changed test files... 1. ['tests/foreman/sys/test_katello_certs_check.py::TestKatelloCertsCheck']
Changed test folders... 0. []
==================================================================================== test session starts =====================================================================================
platform linux -- Python 3.8.6, pytest-6.2.1, py-1.10.0, pluggy-0.13.1
rootdir: /home/okhatavk/Satellite/robottelo, configfile: pyproject.toml
plugins: services-2.2.1, ibutsu-1.13, cov-2.10.1, xdist-2.2.0, mock-3.5.1, forked-1.3.0, picked-0.4.6
collected 13 items                                                                                                                                                                           

tests/foreman/sys/test_katello_certs_check.py .......ssssss                                                                                                                            [100%]

====================================================================================== warnings summary ======================================================================================
env/lib/python3.8/site-packages/botocore/vendored/requests/packages/urllib3/_collections.py:1
env/lib/python3.8/site-packages/botocore/vendored/requests/packages/urllib3/_collections.py:1
  /home/okhatavk/Satellite/robottelo/env/lib/python3.8/site-packages/botocore/vendored/requests/packages/urllib3/_collections.py:1: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated since Python 3.3, and in 3.9 it will stop working
    from collections import Mapping, MutableMapping

-- Docs: https://docs.pytest.org/en/stable/warnings.html
==================================================================== 7 passed, 6 skipped, 2 warnings in 720.43s (0:12:00) ====================================================================
(
```

### CODE REVIEW

File: tests/foreman/data/certs.sh
Comment (swadeley) [inline tests/foreman/data/certs.sh:7]:
```suggestion
  echo "Generating CA"
```

File: tests/foreman/data/certs.sh
Comment (swadeley) [inline tests/foreman/data/certs.sh:16]:
```suggestion
  echo "Generating CA"
```

File: tests/foreman/data/certs.sh
Comment (swadeley) [inline tests/foreman/data/certs.sh:25]:
```suggestion
  echo "Generating CA bundle"
```

File: tests/foreman/data/certs.sh
Comment (swadeley) [inline tests/foreman/data/certs.sh:33]:
```suggestion
  echo "Generating server certificate"
```

File: tests/foreman/data/certs.sh
Comment (swadeley) [inline tests/foreman/data/certs.sh:43]:
```suggestion
  echo "Generating invalid server certificate"
```

File: tests/foreman/data/certs.sh
Comment (swadeley) [inline tests/foreman/data/certs.sh:53]:
```suggestion
  echo "Generating server certificate"
```

File: tests/foreman/data/certs.sh
Comment (swadeley) [inline tests/foreman/data/certs.sh:63]:
```suggestion
  echo "Generating shortname server certificate"
```

File: tests/foreman/sys/test_katello_certs_check.py
Comment (swadeley) [inline tests/foreman/sys/test_katello_certs_check.py:211]: As this test does not generate the certs, would it not be better to say something like:
1. Get valid wildcard certs from `generate_certs`
or:
1. Call `generate_certs` to get valid wildcard certs

File: tests/foreman/sys/test_katello_certs_check.py
Comment (swadeley) [inline tests/foreman/sys/test_katello_certs_check.py:243]: same as above, the shell script is not in this test, consider other wording

**Review by swadeley — APPROVED** [review]
ACK, excellent work, especially creating a real bundle!

File: tests/foreman/data/certs.sh
Comment (devendra104) [inline tests/foreman/data/certs.sh:31]: We can eliminate these duplicate lines by creating a common function.
```
function cert_check(){
   CERT_NAME = $1
   if [[ ! -f "$CERTS_DIR/${CERT_NAME}.key" || ! -f "$CERTS_DIR/$CERT_NAME.crt" ]]; then
        echo "Generating server certificate"
        openssl genrsa -out $CERTS_DIR/$CERT_NAME.key 2048
        openssl req -new -key $CERTS_DIR/$CERT_NAME.key -out $CERTS_DIR/$CERT_NAME.csr -subj "/CN=foreman.example.com"
        openssl x509 -req -in $CERTS_DIR/$CERT_NAME.csr -CA $CERTS_DIR/$CA_CERT_NAME.crt -CAkey $CERTS_DIR/$CA_CERT_NAME.key -CAcreateserial -out $CERTS_DIR/$CERT_NAME.crt -days 3650 -sha256 -extfile extensions.txt -extensions extensions
    else
        echo "Server certificate exists. Skipping."
   fi
}

for cert_name in foreman.example.com  invalid wildcard shortname ; do
    cert_check $cert_name
done
```

Comment (omkarkhatavkar) [reply tests/foreman/data/certs.sh:31]: We also need to check the -subj "/CN=foreman.example.com" which will vary in multiple cases. just not thinking over-complicate this multiple for loops

File: tests/foreman/sys/test_katello_certs_check.py
Comment (mshriver) [inline tests/foreman/sys/test_katello_certs_check.py:254]: Please use f-string

File: tests/foreman/sys/test_katello_certs_check.py
Comment (mshriver) [inline tests/foreman/sys/test_katello_certs_check.py:262]: Would recommend using `result` here, as `each_result` makes for confusing reading.

File: tests/foreman/sys/test_katello_certs_check.py
Comment (mshriver) [inline tests/foreman/sys/test_katello_certs_check.py:268]: I think you should call pytest.fail here

File: tests/foreman/sys/test_katello_certs_check.py
Comment (mshriver) [inline tests/foreman/sys/test_katello_certs_check.py:234]: Why doesn't this test function include use of `certs_cleanup`?

File: tests/foreman/sys/test_katello_certs_check.py
Comment (mshriver) [inline tests/foreman/sys/test_katello_certs_check.py:231]: This doesn't really follow the established pattern of using `robottelo.datafactory` to define invalid inputs that can be parametrized for test functions, any reason for that?  Right now its only used for this one test function, but you defined the data structure at the top of the module, instead of parametrizing the four fields directly.  Are you planning on using invalid_inputs for future test coverage?

File: tests/foreman/sys/test_katello_certs_check.py
Comment (mshriver) [inline tests/foreman/sys/test_katello_certs_check.py:126]: Consider using f-string

**Review by mshriver — CHANGES_REQUESTED** [review]
Couple individual comments to address.

File: tests/foreman/sys/test_katello_certs_check.py
Comment (omkarkhatavkar) [reply tests/foreman/sys/test_katello_certs_check.py:234]: It is handled in the shell file hence no need.

File: tests/foreman/sys/test_katello_certs_check.py
Comment (omkarkhatavkar) [reply tests/foreman/sys/test_katello_certs_check.py:254]: I would have done f-strings it would make code little more ugly and non-understable and based I made it based on the coding standard https://github.com/SatelliteQE/robottelo/blame/master/docs/code_standards.rst#L47. Even you think this can be written little more nicely let me know I can change it.

File: tests/foreman/sys/test_katello_certs_check.py
Comment (omkarkhatavkar) [reply tests/foreman/sys/test_katello_certs_check.py:262]: I just wanted to use the earlier result variable hence make it like that.

File: tests/foreman/sys/test_katello_certs_check.py
Comment (omkarkhatavkar) [reply tests/foreman/sys/test_katello_certs_check.py:231]: Yes, I wanted to use for future test coverage this is raw structure, in case if it required to use multiple test functions across files I can update it to robottelo.datafactory

File: tests/foreman/data/certs.sh
Comment (devendra104) [reply tests/foreman/data/certs.sh:31]: yes, you are right, you can achieve it in this simpler way.
```
function cert_check(){
   CERT_NAME=$1   
   SUBJ=$2
   if [[ ! -f "$CERTS_DIR/${CERT_NAME}.key" || ! -f "$CERTS_DIR/$CERT_NAME.crt" ]]; then
        echo "Generating server certificate"
        openssl genrsa -out $CERTS_DIR/$CERT_NAME.key 2048
        openssl req -new -key $CERTS_DIR/$CERT_NAME.key -out $CERTS_DIR/$CERT_NAME.csr -subj $SUBJ
        openssl x509 -req -in $CERTS_DIR/$CERT_NAME.csr -CA $CERTS_DIR/$CA_CERT_NAME.crt -CAkey $CERTS_DIR/$CA_CERT_NAME.key -CAcreateserial -out $CERTS_DIR/$CERT_NAME.crt -days 3650 -sha256 -extfile extensions.txt -extensions extensions
    else
        echo "Server certificate exists. Skipping."
   fi
}
cert_check "foreman.example.com" "/CN=foreman.example.com"
cert_check "invalid" "/CN=*.example.com"
cert_check "wildcard" "/CN=foreman"
cert_check "shortname" "/CN=Thirdparty CA"


```

File: tests/foreman/sys/test_katello_certs_check.py
Comment (mshriver) [reply tests/foreman/sys/test_katello_certs_check.py:255]: There is certainly room for personal preference here, if you'd like to keep it how it is.

I think its plenty readable even with the dict references, not a blocking opinion though :)

Consider using multi-line strings when writing commands like this.

```
result = connection.run(
    'katello-certs-check '
    f'-c {dict_var["first_key"]} '
    f'-k {dict_var["second_key"]} '
    f'-b {dict_var["third_key"]}'
)
```

File: tests/foreman/sys/test_katello_certs_check.py
Comment (mshriver) [reply tests/foreman/sys/test_katello_certs_check.py:263]: Not sure I follow your explanation here, as you'd still be using `results` as `for result in results`

File: tests/foreman/data/certs.sh
Comment (omkarkhatavkar) [reply tests/foreman/data/certs.sh:31]: okay @devendra104 thanks for the efforts let me try that

File: tests/foreman/sys/test_katello_certs_check.py
Comment (mshriver) [reply tests/foreman/sys/test_katello_certs_check.py:235]: I think its a better pattern to clean up all the certs, regardless of invalid or valid.

How is the script cleanup handled in the shell file?  I don't see any commands there that would remove certificates. The shell file is only loaded and run with generate_certs during test setup.  What happens when a statement within the test function fails?

File: tests/foreman/sys/test_katello_certs_check.py
Comment (omkarkhatavkar) [inline tests/foreman/sys/test_katello_certs_check.py:266]: here I'm using `result.stdout, result.stderr` not wanted to use the same variable

File: tests/foreman/sys/test_katello_certs_check.py
Comment (omkarkhatavkar) [reply tests/foreman/sys/test_katello_certs_check.py:263]: this is want I meant https://github.com/SatelliteQE/robottelo/pull/8297/files#r563774384

File: tests/foreman/sys/test_katello_certs_check.py
Comment (mshriver) [reply tests/foreman/sys/test_katello_certs_check.py:266]: Understood, I think that's even more reason to update the variable naming here to be more clear.

```
certs_check_result = connection.run('katello_certs_check <etc>')
for each result in certs_check_result.stdout.split('\n\n'):
    if error['check'] in result:
        assert '[FAIL]' in result
        assert error['message'] in f'{certs_check_result.stdout} {certs_check_result.stderr}'
```

Comment (omkarkhatavkar) [reply tests/foreman/sys/test_katello_certs_check.py:266]: okay more cleaner, I like it

File: tests/foreman/sys/test_katello_certs_check.py
Comment (mshriver) [inline tests/foreman/sys/test_katello_certs_check.py:226]: Did this change? Why is .format being used to inject hard-coded strings?  Could have sworn in prior commit these were variables.

**Review by mshriver — APPROVED** [review]
Approved, waiting on re-review from @swadeley or @devendra104 before merge.

File: tests/foreman/data/certs.sh
Comment (devendra104) [inline tests/foreman/data/certs.sh:42]: We can also keep the other statements into ```generate_certs``` function

```
function generate_certs(){
   CERT_NAME=$1
   SUBJ=$2
   if [[ ! -f "$CERTS_DIR/${CERT_NAME}.key" || ! -f "$CERTS_DIR/$CERT_NAME.crt" ]]; then
        if [[ $CERT_NAME == "ca-thirdparty" || "$CERT_NAME" == "ca" ]]; then
            echo "Generating $CERT_NAME certificate"
            openssl genrsa -out $CERTS_DIR/$CA_CERT_NAME.key 2048
            openssl req -x509 -new -nodes -key $CERTS_DIR/$CERT_NAME.key -sha256 -days 3650 -out $CERTS_DIR/$CERT_NAME.crt -subj $SUBJ
        elif [[ $CERT_NAME == "ca-bundle" ]]; then
             echo "Generating $CERT_NAME bundle"
             cat $CERTS_DIR/$CERT_NAME.crt $CERTS_DIR/$CERT_NAME.crt > $CERTS_DIR/$CA_BUNDLE.crt
        else
            echo "Generating $CERT_NAME certificate"
            openssl genrsa -out $CERTS_DIR/$CERT_NAME.key 2048
            openssl req -new -key $CERTS_DIR/$CERT_NAME.key -out $CERTS_DIR/$CERT_NAME.csr -subj $SUBJ
            openssl x509 -req -in $CERTS_DIR/$CERT_NAME.csr -CA $CERTS_DIR/$CA_CERT_NAME.crt -CAkey $CERTS_DIR/$CA_CERT_NAME.key -CAcreateserial -out $CERTS_DIR/$CERT_NAME.crt -days 3650 -sha256 -extfile extensions.txt -extensions extensions
        fi
    else
        echo "Server $CERT_NAME certificate exists. Skipping."
   fi
}

```

Comment (omkarkhatavkar) [reply tests/foreman/data/certs.sh:42]: I'm not a fan of multiple nest if-else statements, it increases the complexity and confusions.

**Review by devendra104 — APPROVED** [review]
ACK

**Pull request merged**
